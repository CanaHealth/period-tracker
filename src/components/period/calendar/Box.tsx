import * as React from 'react';

type BoxProps = {
  color: 'normal' | 'flow' | 'ovulation';
  weekDay?: 'Mo' | 'Tu' | 'We' | 'Th' | 'Fr' | 'Sa' | 'Su';
  // prediction: true | false;
} & React.ComponentPropsWithoutRef<'div'>;

export default function Box({ color = 'normal', weekDay }: BoxProps) {
  const colorClasses = {
    normal: 'bg-gray-200',
    flow: 'bg-magenta-94',
    ovulation: 'bg-blue-81',
  };

  return (
    <span
      className={
        'flex h-8 w-8 items-center justify-center rounded-md text-xs ' +
        colorClasses[color]
      }
    >
      {weekDay}
    </span>
  );
}
