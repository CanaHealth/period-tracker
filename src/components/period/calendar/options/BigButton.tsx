import * as React from 'react';

import clsxm from '@/lib/clsxm';

export type BigButtonProps = {
  icon?: React.ReactNode;
  iconLocation?: 't' | 'b' | 'l' | 'r';
  height?: '10' | '20' | '24' | '32';
  text?: string;
  className?: string;
  OnClickDo?: () => void;
};

const BigButton: React.FC<BigButtonProps> = ({
  icon,
  text,
  height = 10,
  iconLocation = 'r',
  OnClickDo,
  className,
}) => {
  return (
    <button
      aria-label={text}
      type='button'
      className={clsxm(
        'mx-auto flex w-44',
        'flex-col items-center justify-center rounded shadow-sm',
        'border-blue-93 rounded-3xl border bg-white  text-indigo-800',
        'transition-all ease-out active:bg-gray-800 active:text-white',
        'hover:bg-gray-600 hover:text-white hover:shadow-md',
        [
          height === '10' && !text && ['h-10'],
          height === '20' && ['h-20'],
          height === '24' && ['h-24'],
          height === '32' && ['h-32'],
          !text && ['text-xl'],
          !icon && ['text-xl'],
        ],
        className
      )}
      onClick={OnClickDo}
    >
      <div
        className={clsxm('flex items-center justify-center text-inherit', [
          iconLocation === 't' && ['flex-col'],
          iconLocation === 'b' && ['flex-col-reverse'],
          iconLocation === 'l' && ['flex-row'],
          iconLocation === 'r' && ['flex-row-reverse'],
        ])}
      >
        {icon && <span className='m-1'>{icon}</span>}
        {text && <span className='m-1 w-max'>{text}</span>}
      </div>
    </button>
  );
};

export default BigButton;
