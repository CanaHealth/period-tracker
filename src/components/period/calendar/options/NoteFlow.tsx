/*
We are working inside extremely well written software repository that included detailed documentation. 
The following information will help us carefully choose the typescript code for the job

A brief description of the behavior the React Functional Component makes possible:
A user selects how heavy her period was, to store the information, so that she can track her cycle.

A summary of what the React Functional Component is:
A modal of three square buttons that pops up after you pick a date to add a note too from a calendar.

A couple words about what the React Functional Component looks like from a user's experience:
A large rectangle that fills the field of view in some creative way. There will be three options, with a custom icon in each.
the one all the way to the left is 'low' with an icon of just half a drop filled in
the middle option is 'average' with the full drop icon colored in.
and the final is high with two drops filled in.
*/

import * as React from 'react';
import { BsDroplet, BsDropletFill, BsDropletHalf } from 'react-icons/bs';

import clsxm from '@/lib/clsxm';

const greyButtonClass = 'bg-gray-200 text-gray-600';

export const defaultProps: NoteFlowProps = {
  flowdata: { howHeavy: 'none', dataID: 'tyv4t4wf4t' },
  buttonProps: [
    {
      icon: <BsDroplet />,
      text: 'low',
    },

    {
      icon: <BsDropletHalf />,
      text: 'average',
    },

    {
      icon: <BsDropletFill />,
      text: 'high',
    },
  ],
};

type FlowData = {
  howHeavy: string;
  dataID: string;
};

export type NoteFlowProps = {
  flowdata: FlowData;

  buttonProps: {
    icon: React.ReactNode;
    text: string;
  }[];
  className?: string;
} & React.ComponentPropsWithoutRef<'div'>;

const NoteFlow: React.FC<NoteFlowProps> = ({
  flowdata,
  buttonProps,
  className,
}) => {
  const [howHeavy, setHowHeavy] = React.useState(flowdata.howHeavy);
  const handleSetHowHeavy = (newHowHeavy: string) => {
    setHowHeavy(newHowHeavy);
  };

  // TODO Handle Save Flow Data to local storage and add a Submit button to the modal

  return (
    <div className={clsxm('', className)}>
      <h1 className='w-full text-2xl font-semibold'>Todays flow: {howHeavy}</h1>
      <div className=' flex flex-col-reverse items-center justify-center md:flex-row'>
        {buttonProps.map((button, index) => (
          <button
            aria-label={button.text}
            key={index}
            type='button'
            className={clsxm(
              'm-2 flex h-44 w-44 flex-col items-center justify-center rounded border-2 border-white shadow-sm hover:shadow-md',

              button.text == howHeavy
                ? 'border border-blue-93'
                : greyButtonClass
            )}
            onClick={() => handleSetHowHeavy(button.text)}
          >
            <div // dynamically set the color
              className={clsxm(
                'flex flex-col items-center justify-center text-center align-middle',
                button.text == howHeavy ? 'text-indigo-600' : 'text-gray-600'
              )}
            >
              {button.icon}
              {button.text}
            </div>
          </button>
        ))}
      </div>
    </div>
  );
};

export default NoteFlow;
