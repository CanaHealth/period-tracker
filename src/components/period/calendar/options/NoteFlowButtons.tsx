import * as React from 'react'

import clsxm from '@/lib/clsxm'

import {
  FlowIntensity,
  NoteFlowProps,
} from '@/components/period/calendar/options/NoteFlow'

export type buttonProp = {
  icon: React.ReactNode
  text: FlowIntensity
  className?: string
}

export type buttonsProps = buttonProp[]

type NoteFlowButtonsProps = {
  renderButton: buttonProp
  flowdata: NoteFlowProps['flowdata']
  index: number
  handleSubmit: (flowdata: NoteFlowProps['flowdata']) => void

  className?: string
} & React.ComponentPropsWithoutRef<'div'>

const NoteFlowButtons: React.FC<NoteFlowButtonsProps> = ({
  flowdata,
  renderButton,
  index,
  className,
  handleSubmit,
}) => {
  // TODO:

  return (
    <button
      aria-label={renderButton.text}
      key={index}
      type='button'
      className={clsxm(
        'm-2 flex h-24 w-44 flex-col items-center justify-center rounded border-2 border-white shadow-sm hover:shadow-md md:h-44',
        className,
        renderButton.text == flowdata.howHeavy
          ? 'border-blue-93 border'
          : 'bg-gray-200 text-gray-600'
      )}
      onClick={() => handleSubmit({ ...flowdata, howHeavy: renderButton.text })}
    >
      <div // dynamically set the color
        className={clsxm(
          'flex flex-col items-center justify-center text-center align-middle',
          renderButton.text == flowdata.howHeavy
            ? 'text-indigo-600'
            : 'text-gray-600'
        )}
      >
        {renderButton.icon}
        {renderButton.text}
      </div>
    </button>
  )
}

export default NoteFlowButtons
