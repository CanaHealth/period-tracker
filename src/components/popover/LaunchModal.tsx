/*

A summary of what it is:
a button that launches a hidden component s a modal


A brief description of the behavior it makes possible:
grouping conceptually linked section of the UI more closely with the current action


A couple words about what it looks like from a user's experience:
I click a button and a model shows up

*/

import * as React from 'react';

import clsxm from '@/lib/clsxm';

import Box from '../period/calendar/Box';

type LaunchModalProps = {
  children: React.ReactNode;
  visible: boolean;
  className?: string;
  buttonOf?: React.ReactNode;
} & React.ComponentPropsWithoutRef<'div'>;

const LaunchModal: React.FC<LaunchModalProps> = ({
  className,
  visible,
  children,
  buttonOf = <Box />,
}) => {
  const [visibleBool, setVisible] = React.useState(visible);
  const modalRef = React.useRef<HTMLDivElement>(null);

  React.useEffect(() => {
    if (visibleBool) {
      modalRef.current?.focus();
    }
  }, [visibleBool]);

  const handleVisibility = () => {
    setVisible(!visibleBool);
  };

  return (
    <div className={clsxm('relative', className)}>
      <button
        aria-label='launch-modal'
        type='button'
        className={clsxm('', {
          hidden: visibleBool,
        })}
        onClick={handleVisibility}
      >
        {buttonOf}
      </button>

      <div
        aria-label='model'
        data-Testid='model'
        role='dialog'
        aria-hidden={!visibleBool}
        className={clsxm(
          'absolute top-1/2 left-1/2 flex -translate-x-1/2 -translate-y-1/2 transform flex-col items-center justify-center space-y-4 rounded border-4 border-white p-16 shadow-md md:p-24',
          { hidden: !visibleBool }
        )}
        ref={modalRef}
      >
        <button
          aria-label='close-modal'
          type='button'
          className={clsxm(
            'absolute top-0 right-0 m-4 rounded-full border-4 border-white bg-gray-99 px-6 py-4 text-black hover:bg-gray-400 hover:text-white',
            {
              hidden: !visibleBool,
            }
          )}
          onClick={handleVisibility}
        >
          close
        </button>
        {children}
      </div>
    </div>
  );
};

export default LaunchModal;
