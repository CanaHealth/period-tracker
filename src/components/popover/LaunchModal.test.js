/* eslint-disable unused-imports/no-unused-imports */
import { describe, expect, it, screen } from '@jest/globals';
import { fireEvent, render } from '@testing-library/react';
import { axe } from 'jest-axe';
import * as React from 'react';

import Box from '@/components/period/calendar/Box';
import { defaultProps } from '@/components/period/calendar/options/NoteFlow';
import NoteFlow from '@/components/period/calendar/options/NoteFlow';
import LaunchModal from '@/components/popover/LaunchModal';

describe('LaunchModal', () => {
  it('is accessible', async () => {
    const { container } = render(
      <LaunchModal visible={false} buttonOf={Box}>
        <NoteFlow {...defaultProps} />
      </LaunchModal>
    );
    expect(await axe(container)).toHaveNoViolations();
  });

  it('is accessible with buttonOf', async () => {
    const { container } = render(
      <LaunchModal visible={false} buttonOf={Box}>
        <NoteFlow {...defaultProps} />
      </LaunchModal>
    );
    expect(await axe(container)).toHaveNoViolations();
  });

  it('should launch the modal when the button is clicked', () => {
    const { getByRole } = render(
      <LaunchModal visible={false} buttonOf={<Box />}>
        <NoteFlow {...defaultProps} />
      </LaunchModal>
    );
    const button = getByRole('button', { name: 'launch-modal' });
    fireEvent.click(button);

    const modal = getByRole('dialog');
    expect(modal).not.toHaveClass('hidden');
  });

  it('should close the modal when the button is clicked and the modal is currently open', () => {
    const { getByRole, getByLabelText } = render(
      <LaunchModal visible={true} buttonOf={<Box />}>
        <NoteFlow {...defaultProps} />
      </LaunchModal>
    );
    const button = getByRole('button', { name: 'launch-modal' });
    fireEvent.click(button);

    const modal = getByLabelText('model');
    expect(modal).toHaveClass('hidden');
  });
});
