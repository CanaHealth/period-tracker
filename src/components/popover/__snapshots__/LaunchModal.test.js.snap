// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LaunchModal should render the same 1`] = `
<div>
  <div
    class="relative"
  >
    <button
      aria-label="launch-modal"
      class=""
      type="button"
    >
      <span
        class="relative flex h-8 w-8 items-center justify-center rounded-md text-xs bg-gray-200 text-gray-700"
      />
    </button>
    <div
      aria-hidden="true"
      aria-label="model"
      class="absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 transform flex-col items-center justify-center space-y-4 rounded border-4 border-white p-16 shadow-md md:p-24 hidden"
      data-testid="model"
      role="dialog"
    >
      <button
        aria-label="close-modal"
        class="absolute top-0 right-0 m-4 rounded-full border-4 border-white bg-gray-99 px-6 py-4 text-black hover:bg-gray-400 hover:text-white hidden"
        type="button"
      >
        close
      </button>
    </div>
  </div>
</div>
`;
